components/zcash_address/src/kind/unified.rs

```diff
@@ -24,32 +28,156 @@ pub(crate) const REGTEST: &str = "uregtest";
+            (Self::Orchard, _) => cmp::Ordering::Less,
+            (_, Self::Orchard) => cmp::Ordering::Greater,
+
+            (Self::Sapling, _) => cmp::Ordering::Less,
+            (_, Self::Sapling) => cmp::Ordering::Greater,
+
+            (Self::P2sh, _) => cmp::Ordering::Less,
+            (_, Self::P2sh) => cmp::Ordering::Greater,
+
+            (Self::P2pkh, _) => cmp::Ordering::Less,
+            (_, Self::P2pkh) => cmp::Ordering::Greater,
+        }
+    }
+}
+
+impl PartialOrd for Typecode {
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl From<u8> for Typecode {
+    fn from(typecode: u8) -> Self {
+        match typecode {
+            0x00 => Typecode::P2pkh,
+            0x01 => Typecode::P2sh,
+            0x02 => Typecode::Sapling,
+            0x03 => Typecode::Orchard,
+            _ => Typecode::Unknown(typecode),
+        }
+    }
+}
+
+impl From<Typecode> for u8 {
+    fn from(t: Typecode) -> Self {
+        match t {
+            Typecode::P2pkh => 0x00,
+            Typecode::P2sh => 0x01,
+            Typecode::Sapling => 0x02,
+            Typecode::Orchard => 0x03,
+            Typecode::Unknown(typecode) => typecode,
+        }
+    }
+}
+
+impl Typecode {
+    fn is_shielded(&self) -> bool {
+        match self {
+            Typecode::P2pkh | Typecode::P2sh => false,
+            // Assume that unknown typecodes are shielded, because they might be.
+            _ => true,
+        }
+    }
+}
+
+/// An error while attempting to parse a string as a Zcash address.
+#[derive(Debug, PartialEq)]
+pub enum ParseError {
+    /// The unified address contains both P2PKH and P2SH receivers.
+    BothP2phkAndP2sh,
+    /// The unified address contains a duplicated typecode.
+    DuplicateTypecode(Typecode),
+    /// The string is an invalid encoding.
+    InvalidEncoding,
+    /// The unified address only contains transparent receivers.
+    OnlyTransparent,
+}
+
+impl fmt::Display for ParseError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ParseError::BothP2phkAndP2sh => write!(f, "UA contains both P2PKH and P2SH receivers"),
+            ParseError::DuplicateTypecode(typecode) => {
+                write!(f, "Duplicate typecode {}", u8::from(*typecode))
+            }
+            ParseError::InvalidEncoding => write!(f, "Invalid encoding"),
+            ParseError::OnlyTransparent => write!(f, "UA only contains transparent receivers"),
+        }
+    }
+}
+
+impl Error for ParseError {}
+
 /// The set of known Receivers for Unified Addresses.
 ///
 /// This enum is an internal-only type, and is maintained in preference order, so that the
 /// derived [`PartialOrd`] will sort receivers correctly. From its documentation:
 ///
 /// > When derived on enums, variants are ordered by their top-to-bottom discriminant
 /// > order.
```

Comment:
Given that `cmp` is implemented explicitly, wouldn't it be clearer to not derive `PartialEq` and use
```
impl PartialOrd for Person {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}
``` (as suggested in the [PartialOrd documentation](https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html))? That would remove the risk of inconsistent implementations.

Reply:
Oh, you do have that impl. But now I'm confused as to why the comment is relevant.

Reply:
The comment was previously relevant, but after implementing receiver preference ordering in C++, and now needing `Recipient` to be a public type, I decided to make the ordering explicit (and identical to what I'm doing in C++), and removing any dependency on how Rust implements ordering.

I'll push a fix to update the comment.